no_proxy for winrm

- Packer version from 1.1.3
- Host platform: Linux and Windows
- Guest: Windows
- Waiting to connect to WinRM...
   Sometimes it connects (see below), sometimes it throws back a Proxy error. 

ssh_interface (private) helps a little but still causes problems.

The issue I have is that we do not use the AWS DNS so adding .internal to our no_proxy doesn't help. However I can work around this by having a script update my hosts file with the IP/FQDN when using private_dns with the ssh_interface flag. While using the no_proxy=".internal", this works for the most part but it's not 100% reliable.  

When using private_ip Adding every IP to no_proxy doesn't help either as sometimes the IP given is at the end of the range which doesn't appear to work - probably due to an environment variable limitation. 

Is it possible to have packer automatically add the IP to no_proxy or append to it with the Private IP it gets as part of the boot sequence?

Thoughts?

Unfortuantly no luck on my side.

Packer v1.1.4-dev
            "winrm_no_proxy": true,
```
WinRMInsecure:true WinRMNoProxy:true WinRMPassword: WinRMPort:5986 WinRMTimeout:30m0s WinRMUseNTLM:false WinRMUseSSL:true WinRMUser:Administrator

2018/02/08 11:45:13 packer: 2018/02/08 11:45:13 [INFO] Attempting WinRM connection...
2018/02/08 11:45:13 packer: 2018/02/08 11:45:13 [DEBUG] connecting to remote shell using WinRM
2018/02/08 11:46:28 packer: 2018/02/08 11:46:28 [ERROR] connection error: http response error: 503 - invalid content type
2018/02/08 11:46:28 packer: 2018/02/08 11:46:28 [ERROR] WinRM connection err: http response error: 503 - invalid content type
```

--- a/helper/communicator/config.go
+++ b/helper/communicator/config.go
@@ -41,14 +41,15 @@ type Config struct {
 	SSHReadWriteTimeout       time.Duration `mapstructure:"ssh_read_write_timeout"`
 
 	// WinRM
-	WinRMUser               string        `mapstructure:"winrm_username"`
-	WinRMPassword           string        `mapstructure:"winrm_password"`
 	WinRMHost               string        `mapstructure:"winrm_host"`
+	WinRMInsecure           bool          `mapstructure:"winrm_insecure"`
+	WinRMNoProxy            bool          `mapstructure:"winrm_no_proxy"`
+	WinRMPassword           string        `mapstructure:"winrm_password"`
 	WinRMPort               int           `mapstructure:"winrm_port"`
 	WinRMTimeout            time.Duration `mapstructure:"winrm_timeout"`
-	WinRMUseSSL             bool          `mapstructure:"winrm_use_ssl"`
-	WinRMInsecure           bool          `mapstructure:"winrm_insecure"`
 	WinRMUseNTLM            bool          `mapstructure:"winrm_use_ntlm"`
+	WinRMUseSSL             bool          `mapstructure:"winrm_use_ssl"`
+	WinRMUser               string        `mapstructure:"winrm_username"`
 	WinRMTransportDecorator func() winrm.Transporter
 }
 

--- a/helper/communicator/step_connect_test.go
+++ b/helper/communicator/step_connect_test.go
@@ -3,10 +3,12 @@ package communicator
 import (
 	"bytes"
 	"context"
+	"os"
 	"testing"
 
 	"github.com/hashicorp/packer/helper/multistep"
 	"github.com/hashicorp/packer/packer"
+	"github.com/stretchr/testify/assert"
 )
 
 func TestStepConnect_impl(t *testing.T) {
@@ -26,6 +28,28 @@ func TestStepConnect_none(t *testing.T) {
 	// run the step
 	if action := step.Run(context.Background(), state); action != multistep.ActionContinue {
 		t.Fatalf("bad action: %#v", action)
+	}
+}
+
+var noProxyTests = []struct {
+	current  string
+	expected string
+}{
+	{"", "foo:1"},
+	{"foo:1", "foo:1"},
+	{"foo:1,bar:2", "foo:1,bar:2"},
+	{"bar:2", "bar:2,foo:1"},
+}
+
+func TestStepConnect_setNoProxy(t *testing.T) {
+	key := "NO_PROXY"
+	for _, tt := range noProxyTests {
+		if value := os.Getenv(key); value != "" {
+			os.Unsetenv(key)
+			defer func() { os.Setenv(key, value) }()
+			os.Setenv(key, tt.current)
+			assert.Equal(t, tt.expected, os.Getenv(key), "env not set correctly.")
+		}
 	}
 }
 

--- a/helper/communicator/step_connect_winrm.go
+++ b/helper/communicator/step_connect_winrm.go
@@ -7,6 +7,7 @@ import (
 	"fmt"
 	"io"
 	"log"
+	"os"
 	"strings"
 	"time"
 
@@ -128,6 +129,12 @@ func (s *StepConnectWinRM) waitForWinRM(state multistep.StateBag, cancel <-chan 
 			}
 		}
 
+		if s.Config.WinRMNoProxy {
+			if err := setNoProxy(host, port); err != nil {
+				return nil, fmt.Errorf("Error setting no_proxy: %s", err)
+			}
+		}
+
 		log.Println("[INFO] Attempting WinRM connection...")
 		comm, err = winrm.New(&winrm.Config{
 			Host:               host,
@@ -181,4 +188,23 @@ func (s *StepConnectWinRM) waitForWinRM(state multistep.StateBag, cancel <-chan 
 	}
 
 	return comm, nil
+}
+
+// setNoProxy configures the $NO_PROXY env var
+func setNoProxy(host string, port int) error {
+	current := os.Getenv("NO_PROXY")
+	p := fmt.Sprintf("%s:%d", host, port)
+	// not set
+	//	set
+	// is set and not contains
+	//	set
+	// is set and contains
+	if current == "" {
+		return os.Setenv("NO_PROXY", p)
+	}
+	if !strings.Contains(current, p) {
+		return os.Setenv("NO_PROXY", strings.Join([]string{current, p}, ","))
+	}
+	return nil
+
 }

--- a/website/source/docs/templates/communicator.html.md
+++ b/website/source/docs/templates/communicator.html.md
@@ -139,16 +139,21 @@ The WinRM communicator has the following options.
 
 -   `winrm_password` (string) - The password to use to connect to WinRM.
 
+-   `winrm_insecure` (boolean) - If true, do not check server certificate
+    chain and host name
+
+*   `winrm_no_proxy` (boolean) - Setting this to `true` adds the remote
+    `host:post` to the `NO_PROXY` environment variable. This has the effect of
+    bypassing any configured proxies when connecting to the remote host.
+    Default to `false`.
+
 -   `winrm_timeout` (string) - The amount of time to wait for WinRM to
     become available. This defaults to "30m" since setting up a Windows
     machine generally takes a long time.
-
--   `winrm_use_ssl` (boolean) - If true, use HTTPS for WinRM
-
--   `winrm_insecure` (boolean) - If true, do not check server certificate
-    chain and host name
 
 -   `winrm_use_ntlm` (boolean) - If true, NTLM authentication will be used for WinRM,
     rather than default (basic authentication), removing the requirement for basic
     authentication to be enabled within the target guest. Further reading for remote
     connection authentication can be found [here](https://msdn.microsoft.com/en-us/library/aa384295(v=vs.85).aspx).
+
+-   `winrm_use_ssl` (boolean) - If true, use HTTPS for WinRM

